### DISCUSSION TITLE
Replacing the Twye1s chip with D1 mini

### DISCUSSION THREAD

**Discussion by jazzmonger** [jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip#23] [General]
I ran into a nasty snag at some point and the tuya chip would randomly reboot. The only way to see why is to hook up the serial port for monitoring.  No matter what I tried I could not get that to work. You should see the same thing over serial as you do Wifi except for low level stuff and things the Wifi log misses. I was getting nothing.  After looong discussions on Discord, we felt the chip was malfunctioning. So I decided to try removing it and replacing it with a D1 mini. I used a heat gun to melt the solder and carefully pried it off the pcb, working on small sections at a time. 

Amazingly, it worked. The only connections required to the old chip traces were RX &TX, +5v and gnd.

![BBD8BDD2-27A4-464E-ABBD-F2867F805236](https://user-images.githubusercontent.com/52110065/210706922-a6187ba5-0422-4655-b302-d76367c78cbf.jpeg)
The GPIOs don't do anything anyway so I had a feeling it would work.

After replacing it, the serial log worked fine and It turns out that the random rebooting error was "OOM". Out of memory. A little sleuth work led me to figuring out that  I had the web server enabled which caused the random reboots. Disabling the web server solved it!  

The *Only* way I could actually see this was unsoldering the Tuya chip and putting in the D1 mini. I couldn't get the USB log to display on the Tuya chip no matter what I tried. 

This turned out to be a completely unnecessary step for this project, but I was at a total impasse and had to try something.  I have several of these display units so this will just become my low level Dev system and I'll put a stock unit back in my Pellet Stove.

**Comment by jazzmonger** [comment]
Not so fast.... decoding the data stream coming from the control module so I could see the temps updated in Real-time proved fruitless.  Using 2 uarts on the Tywe1s didn't work and after hours (days) of staring at HEX codes in excel I ultimately couldn't find a way to decode the data stream.

So, new line of attack. I'm going to tap the line of the exhaust thermocouple and read the voltage to determine the temp. Then I can use this D1 mini I replaced the TYWE1S with which gives me an A0 analog line input. There isn't one on the tywe1s. So it wasn't wasted effort afterall. That will only take 1 more wire from the control to the display.

Continued here https://github.com/jazzmonger/wood-stove-with-TYWE1S-Tuya-chip/discussions/30

**Comment by jwaz73** [comment]
Hey Friend!  I'm blowing the dust off my pellet stove project.  Work and family time have definitely taken their rightful priority lately (new grandson arriving any day now!).  I'm amazed to see the progress you have made over the last few months.

I have been unsuccessful in getting my hands on a spare WiFi module.  I've had my stove manufacturer send me two different display boards but neither came with a WiFi module.

What would be the possibility of replacing the manufacturer's WiFi board completely with a D1 Mini or even an ESP32-based board?  I know the pin alignment won't match but if all that is actually needed is the RX, TX, VCC, and GND, those could be connected with jumpers from the D1 Mini to the 8-pin sockets on the display board.  Looking at the Tuya WiFi module, I don't notice any components other than those found on any standard ESP8266 dev board... Resistors, capacitors, a couple of transistors, a voltage regulator, an LED, and a button switch.  If fact, the only real difference I see between the Tuya board and the D1 Mini is the USB to serial UART and the auto-programming mode logic.

In your work on this have you noticed any display or MCU dependencies on the specific N12210 board?

**Reply by jazzmonger** [reply]
Totally doable, but all the levels from the MCU are 5v, so you have to adjust for that.  I literally unsoldered the Tuya chip on mine and used RX,TX +5V and gnd to the D1 mini+ my code.  That's what I'm doing now, especially since the Tuya chip doesn't have an A0 analog input, and I need that to get real-time measurements from the exhaust thermocouple to better control the burn pot instead of the once every 60s provided by TuyaMCU. The stove burns for 1-2 days on ultra low, but occasionally goes out when the timing is just right sine the ex pipe temp only updates every 60s.

I kept the board the Tuya chip was on and just attached the D1 to that. It actually fits nicely underneath the board.  You could get rid of the Tuya board if u use level adjusters I suppose, although the ones I got are flakey. https://www.amazon.com/gp/product/B07YZTW2SM/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1
Btw, That's what another fellow had in his stove he bought in the Netherlands. Literally a D1 mini board.  Check it out!

https://github.com/jazzmonger/wood-stove-with-TYWE1S-Tuya-chip/discussions/31#discussioncomment-4921035

I'm 90% done with my latest set of mods and just waiting for some USB jacks to get here Fri so I can clean up the connection between the MCU and display board.  I'll post all that probably early next week. It's clean, and a 4 wire usb cable from the MCU to the display is all that's needed for the new signaling controls.

**Reply by jazzmonger** [reply]
> In your work on this have you noticed any display or MCU dependencies on the specific N12210 board?

None whatsoever.

Check out the final build.  Works great.
https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/discussions/32
